clang/docs/SYCLFERFCIntegrationHeaderFooter.md

```diff
@@ -0,0 +1,179 @@
+       id<1> accId,           // arg4 of accessor init function
+       sampler_t smpData      // arg1 of sampler init function
+    )
+    {
+        // Local capture object
+        struct FuncObj local;
+
+        // Reassemble capture object from parts
+        local.i = i;
+        local.s = s;
+        // Call acc1 accessor's init function
+        sycl::accessor::init(&local.acc1, accData, accR1, accR2, accId);
+        // Call smplr sampler's init function
+        sycl::accessor::init(&local.smplr, smpData);
+
+        // Call the kernel body
+        callee(&local);
+    }
+
+    spir_func void callee(struct FuncObj* this)
+    {
+        // body of the kernel invocation
+    }
+```
+
+For details of this transformation, see 'Lowering of SYCL-Kernel'. // TODO: Add link
+
+The integration header then describes the fields of the kernel objects as follows:
+
+```
+namespace sycl {
+  inline namespace _V1 {
+    namespace detail {
+      // names of all kernels defined in the corresponding source
+      static constexpr
+      const char* const kernel_names[] = {
+        "_ZTSZ4mainE12first_kernel"  // mangled name associated with first_kernel
+                                     // demangles to 'typeinfo name for main::first_kernel'
+
+      };
+      // array representing signatures of all kernels defined in the
+      // corresponding source
+      static constexpr
+      const kernel_param_desc_t kernel_signatures[] = {
+        //--- _ZTSZ4mainE12first_kernel
+        { kernel_param_kind_t::kind_std_layout, 4, 0 },
+        { kernel_param_kind_t::kind_std_layout, 8, 4 },
+        { kernel_param_kind_t::kind_accessor, 4062, 12 },
+        { kernel_param_kind_t::kind_sampler, 8, 24 },
+
+        { kernel_param_kind_t::kind_invalid, -987654321, -987654321 },
+      };
+    } // namespace detail
+  } // namespace _V1
+} // namespace sycl
+```
+
+For each kernel, a mangled name associated with it, is stored in the kernel_names array.
+
+And for each kernel, each of its parameter is described in the kernel_signatures array.   The description consists of the following parts:
+
+* an encoding of the type that is captured
+* its size and its offset
+
+In the above example, the first two parameters are standard layout types (the int 'i' and the struct 'test_s'), the latter two though standard layout types are types which need special handling - accessor and sampler respectively.   The last entry acts as a terminating entry.
+
+This allows the run-time library to recreate the function object on the device.
+
+The full set of encoding is:
+
+* kind_accessor  for sycl accessor types
+* kind_std_layout for standard layout types
+* kind_sampler for sycl sampler types
+* kind_pointer for pointer types
+* kind_specialization_constants_buffer for initializing specialization constants
+* kind_stream for sycl stream types and
+* kind_invalid used to indicate the terminal value in the signatures array.
+
```

Comment:
I can't recall off the top of my head if this is the only information in the integration header. Do we just have kernel names and signature? For some reason I thought we had more information there.

Reply:
I wanted to leave out the other details.   I did not feel like it added much.